### ISSUE TITLE
CA1804: Remove unused locals - Doesn't Work

### ISSUE THREAD

**Issue by MelbourneDeveloper** [dotnet/roslyn-analyzers#2109]
#### Analyzer package

Microsoft.CodeAnalysis.FxCopAnalyzers
Microsoft.CodeQuality.Analyzers

This is probably the most fundamental code rule there is. Have I missed a NuGet package or something? If the original code rule isn't going to be implemented, there needs to be some kind of replacement.

#### Package Version

Example: v2.6.3

#### Diagnostic ID

CA1804: Remove unused locals

#### Repro steps

```cs
namespace ClassLibrary1
{
    public static class Class1
    {
        public static void test()
        {
            var asdasd = new object();
        }
    }
}
```
![image](https://user-images.githubusercontent.com/16697547/52613156-f802dd00-2ee0-11e9-897a-76520091bbe8.png)

#### Expected behavior

An error/warning that says 

> "{0} declares a variable, {1}, of type {2}, which is never used or is only assigned to. Use this variable or remove it."

#### Actual behavior

No warning/error

**Comment by mavasani** [comment]
This has been ported over to Visual Studio 2019 as `IDE0059` (message severity by default), which shows up by default in the IDE while typing. You can try this out at https://docs.microsoft.com/en-us/visualstudio/releases/2019/release-notes-preview.

In future, Visual Studio IDE analyzers will be available as a NuGet package to enforce these on CI. At present, we are not going to duplicate the rule in both the FxCop analyzers package and VS2019 IDE.

**Comment by MelbourneDeveloper** [comment]
@mavasani this rule has very strange behavior. I realized just now that if the variable is a string, it will kick in. If it's an object, it will not kick in.

Fine:
![image](https://user-images.githubusercontent.com/16697547/53278282-83d7fd00-375c-11e9-89ec-b52c4649c4f9.png)

Note fine:
![image](https://user-images.githubusercontent.com/16697547/53278297-90f4ec00-375c-11e9-99dc-a142e6f8bbd4.png)

**Comment by mavasani** [comment]
@MelbourneDeveloper You probably again confused with CSxxxx error here - this is the one from the compiler which only flags locals assigned to _constant_ values as unused.

I actually meant that the VS IDE (not the compiler) now generates IDE0059 (see the 0 of 2 messages in the error list), and also fades out unused local. This diagnostic will not be generated during build OR break your build, it is just a live inbuilt analysis in Visual Studio.

**Comment by MelbourneDeveloper** [comment]
@mavasani OK. Thanks

**Comment by mavasani** [comment]
Also note that you need at least VS2019 Preview2 to see IDE0059 violations.

**Comment by wizofaus** [comment]
So how can we ensure that compilations if there are unused local variables? (I actually want compilation failures for ANY unreferenced code/variables/properies etc., or code only called from tests, like I can get from Eclipse for java - it's also smart enough to recognise web controller endpoint functions).

**Comment by mavasani** [comment]
Tagging @sharwell. Once we move all or most of the IDE analyzers to a NuGet package, you should be able to enforce them on the build.

**Comment by MelbourneDeveloper** [comment]
@mavasani  @sharwell this is a major one for me. In fact, I would say that this is probably the most fundamental code rule of all. If variables or members are declared anywhere that are not being used, they should be the first things to get deleted from the code base.

**Comment by mavasani** [comment]
@MelbourneDeveloper I am going to mark this as external as we have already a rule implementation in VS IDE, which will be used to detect these cases and offer a code fix to remove them

**Comment by RobSiklos** [comment]
@mavasani Is there any timeline for moving the IDE analyzers to a NuGet package? Is there any project or something that I can subscribe to in order to track the progress of this? It's a pretty bad step backwards to lose these type of warnings in CI now that we're moving away from the legacy FxCop.

**Comment by mavasani** [comment]
> Is there any timeline for moving the IDE analyzers to a NuGet package? Is there any project or something that I can subscribe to in order to track the progress of this?

Tagging @sharwell @jinujoseph

**Comment by yitzhaks** [comment]
A kind ping.

**Comment by mavasani** [comment]
See update here: https://github.com/dotnet/roslyn/issues/33558#issuecomment-605974524. Adding a NuGet package reference to the CodeStyle NuGet package should enforce IDE0059 to flag all redundant locals and unnecessary assignments to locals/parameters on build/CI.

**Comment by MelbourneDeveloper** [comment]
@mavasani this seems to be a problem in the latest NuGet package Microsoft.CodeAnalysis.NetAnalyzers 5.0.3 still

**Comment by arphox** [comment]
I tried to make [CA1804: Remove unused locals](https://learn.microsoft.com/en-us/visualstudio/code-quality/ca1804?view=vs-2022&tabs=csharp) work for the example on the documentation site (`string[] unusedArray = Environment.GetLogicalDrives();`), but it just doesn't work.

I created a new empty .NET 8 project in VisualStudio to test it.
csproj:
```cs
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
  </PropertyGroup>
</Project>
```

Program.cs:
```cs
using System;
namespace torolheto0001;

public class Program
{
    public static void Main(string[] args)
    {
        string[] unusedArray = Environment.GetLogicalDrives();
    }
    private void UnusedMethod()
    {
        Console.WriteLine("Hello, World!");
    }
}
```
.editorconfig:
```
[*.cs]
# CA1804: Remove unused locals
dotnet_diagnostic.CA1804.severity = warning

# CA1822: Make member static
dotnet_diagnostic.ca1822.severity = warning
dotnet_code_quality.CA1822.api_surface = private
```

CA1822 is included for sanity checking (to test whether editorconfig works).

Build result: CA1822 is reported, CA1804 is not.
![kép](https://github.com/dotnet/roslyn-analyzers/assets/17320094/1e04e853-63e2-4e3f-b8b9-ca4d422ef825)

As my understanding, this is a not working rule.
Could you please share your input on this?

@mavasani

**Comment by lhfiedler** [comment]
IDE0059 is not a replacement for CA1804.  IDE0059 mainly checks is for unneeded assignments to local variables.  If you have a local variable that is the return value of the function call, and you don't use that variable,  IDE0059 won't pick it up.  CA1804 has a found a lot of dead code and bugs for us.

**Comment by SaebAmini** [comment]
I can't get this to work with any of the existing options either as @arphox has mentioned. I've created #7412 for it.

**Comment by areevesat** [comment]
I created issue 81742 in roslyn: https://github.com/dotnet/roslyn/issues/81742 for IDE0059.

Which appears to be a duplicate of https://github.com/dotnet/roslyn/issues/65220, which has been open since 2022.

As far as I can tell, there is still no reliable way to detect unused local variables.

**Comment by lhfiedler** [comment]
We went back to using FxCop so we can have CA1804.  We were running rosalyn / IDE0059  for awhile and it missed so many things .....

**Comment by areevesat** [comment]
I also created issue https://github.com/dotnet/roslyn/issues/81743 to track another rule, CS0219, which ought to capture these but only works for values that are known at compile time.  I have a proposed change in that issue that would catch more unused values.
